File: packages/dd-trace/test/llmobs/parity/scenarios/genai/_runner.js
'use strict'

const path = require('node:path')
const { finish, loadTracer } = require('../common')

const ROOT = path.resolve(__dirname, '../../../../../../../')
const sdkPackage = path.join(ROOT, 'versions/@google/genai')

async function run (scenario) {
  const tracer = loadTracer('google-genai')
  const { GoogleGenAI } = require(sdkPackage).get()
  const client = new GoogleGenAI({
    apiKey: 'test',
    httpOptions: { baseUrl: process.env.PROVIDER_BASE_URL },
  })
  const model = scenario === 'generate-content-model-path' ? 'models/gemini-2.5-flash' : 'gemini-2.5-flash'
  try {
    if (scenario === 'embed-content') {
      await client.models.embedContent({
        model: 'text-embedding-004',
        contents: 'Hello, world!',
        config: { taskType: 'RETRIEVAL_QUERY', outputDimensionality: 8 },
      })
    } else if (scenario === 'generate-content-stream') {
      const stream = await client.models.generateContentStream({
        model,
        contents: 'Hello, world!',
        config: { temperature: 0.5, maxOutputTokens: 100, systemInstruction: 'You are a parity bot.' },
      })
      for await (const chunk of stream) void chunk
    } else {
      await client.models.generateContent({
        model,
        contents: scenario === 'generate-content-tools'
          ? [
              { role: 'user', parts: [{ text: 'Use the tool.' }] },
              {
                role: 'model',
                parts: [{ functionCall: { id: 'call-1', name: 'lookup', args: { value: 'parity' } } }],
              },
              {
                role: 'user',
                parts: [{ functionResponse: { id: 'call-1', name: 'lookup', response: { value: 'ok' } } }],
              },
            ]
          : 'Hello, world!',
        config: scenario === 'generate-content-reasoning'
          ? { temperature: 0.5, maxOutputTokens: 100, systemInstruction: 'You are a parity bot.' }
          : scenario === 'generate-content-tools'
            ? {
                temperature: 0.5,
                maxOutputTokens: 100,
                systemInstruction: 'You are a parity bot.',
                tools: [{
                  functionDeclarations: [{
                    name: 'lookup',
                    description: 'Look up a value.',
                    parameters: { type: 'OBJECT', properties: { value: { type: 'STRING' } } },
                  }],
                }],
              }
            : { temperature: 0.5, maxOutputTokens: 100, systemInstruction: 'You are a parity bot.' },
      })
    }
  } catch (error) {
    void error
  } finally {
    await finish(tracer)
  }
}

module.exports = { run }

File: packages/dd-trace/test/llmobs/parity/scenarios/genai/_runner.py
import os

from google import genai
from google.genai import types

from common import finish, start


def run(scenario):
    start("google_genai")
    client = genai.Client(api_key="test", http_options={"base_url": os.environ["PROVIDER_BASE_URL"]})
    model = "models/gemini-2.5-flash" if scenario == "generate-content-model-path" else "gemini-2.5-flash"
    try:
        if scenario == "embed-content":
            client.models.embed_content(
                model="text-embedding-004",
                contents="Hello, world!",
                config=types.EmbedContentConfig(task_type="RETRIEVAL_QUERY", output_dimensionality=8),
            )
        elif scenario == "generate-content-stream":
            for _ in client.models.generate_content_stream(
                model=model,
                contents="Hello, world!",
                config=types.GenerateContentConfig(
                    temperature=0.5,
                    max_output_tokens=100,
                    system_instruction="You are a parity bot.",
                ),
            ):
                pass
        else:
            contents = "Hello, world!"
            config = types.GenerateContentConfig(
                temperature=0.5,
                max_output_tokens=100,
                system_instruction="You are a parity bot.",
            )
            if scenario == "generate-content-tools":
                contents = [
                    types.Content(role="user", parts=[types.Part.from_text(text="Use the tool.")]),
                    types.Content(
                        role="model",
                        parts=[types.Part(function_call=types.FunctionCall(name="lookup", args={"value": "parity"}, id="call-1"))],
                    ),
                    types.Content(
                        role="user",
                        parts=[types.Part(function_response=types.FunctionResponse(name="lookup", response={"value": "ok"}, id="call-1"))],
                    ),
                ]
                config.tools = [types.Tool(function_declarations=[types.FunctionDeclaration(
                    name="lookup",
                    description="Look up a value.",
                    parameters={"type": "OBJECT", "properties": {"value": {"type": "STRING"}}},
                )])]
            client.models.generate_content(model=model, contents=contents, config=config)
    except Exception:
        pass
    finally:
        finish()
